internal/mode/static/state/graph/service.go

```diff
@@ -0,0 +1,46 @@
+package graph
+
+import (
+	"k8s.io/apimachinery/pkg/types"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+func buildReferencedServicesNames(
```

Comment:
we only care about HTTPRoutes that are part of the Graph -- https://github.com/nginxinc/nginx-gateway-fabric/blob/40f46d34b93c3f9f1e266405e5524e411b3b91d5/internal/mode/static/state/graph/graph.go#L44 ( not all HTTPRoutes in the cluster ) and the ones attached to NGF Gateway,

Note that there are could be HTTPRoutes that do not belong to our Gateway -- we don't want to trigger Graph rebuild for changes to Endpoints/Services for those HTTPRoutes.

The ones that belong to our Gateway will have populated ParentRef https://github.com/nginxinc/nginx-gateway-fabric/blob/40f46d34b93c3f9f1e266405e5524e411b3b91d5/internal/mode/static/state/graph/httproute.go#L62 
More over, attachment.Attached needs to be true. Otherwise, we don't generate NGINX configuration for that route.
We can also exclude Routes which are Invalid, since we don't generate NGINX configuration for such routes.

Note that we don't need to parse HTTPRoute to extract Services again  since we already parse rules https://github.com/nginxinc/nginx-gateway-fabric/blob/40f46d34b93c3f9f1e266405e5524e411b3b91d5/internal/mode/static/state/graph/httproute.go#L67 and build backend refs -- https://github.com/nginxinc/nginx-gateway-fabric/blob/40f46d34b93c3f9f1e266405e5524e411b3b91d5/internal/mode/static/state/graph/httproute.go#L23-L32

so I think we can try to extract service names from there -- https://github.com/nginxinc/nginx-gateway-fabric/blob/40f46d34b93c3f9f1e266405e5524e411b3b91d5/internal/mode/static/state/graph/backend_refs.go#L18

the only problem is that if Service doesn't exist, this field is nil -- https://github.com/nginxinc/nginx-gateway-fabric/blob/40f46d34b93c3f9f1e266405e5524e411b3b91d5/internal/mode/static/state/graph/backend_refs.go#L18 . However, we still want to track non-existing referenced services.  For that to work, I think we can change the logic here https://github.com/nginxinc/nginx-gateway-fabric/blob/40f46d34b93c3f9f1e266405e5524e411b3b91d5/internal/mode/static/state/graph/backend_refs.go#L121-L123 so that we still capture (save) referenced Service ns/name or refactor BackendRef so it doesn't reference a Service object, but rather only stores ns/name.

@kate-osborn  I wonder what you think

Reply:
This makes sense. I talked offline with @bjee19, and we think the best approach is to refactor BackendRef to store the Service name. The Service Resolver will also need to be refactored to resolve endpoints given a Service nsname and a ServicePort. This will take some additional time because there are a fair amount of unit tests that will need to be changed.

Reply:
Alright changes are in! If I could get some eyes on the changes made in `backend_refs.go` and `resolver.go`, that'd be great! Also any thoughts on me storing the ServiceNames on the `Route` would be good! Mostly just updated the tests to work with the changes so no additional tests have been added. (`change_processor_test.go` is heavily WIP)